[WIP] docs: Add design document (#11)
kiagnose/kiagnose: docs/kiagnose_design.md

behavior.

To simplify cluster operation and maintenance, it will be beneficial to allow cluster administrators to validate that
the cluster can support a specific functionality/service and to provide a simple indication that they are working as
intended.

Authors of features and services should be able to easily write and publish predefined checkups that will verify that
their feature / service works as expected on a pre-deployed cluster.

## Goals

- Deploy a checkup to a cluster.
- Run the predefined checkup on a cluster on an existing or on ephemeral Namespace.
- Provide a reference implementation of a checkup.
- Do not disrupt or interfere with existing cluster workloads. This includes making sure no leftovers exist on
  termination.
- Checkup execution should require minimal dependencies, allowing a cluster administrator or some management tool to
  trigger it.
- Deploy and run the checkups on any K8s-based cluster.
- Provide a well-defined API, allowing management components to interact with Kiagnose.

## Non Goals

- Defining content of the checkup, including its correctness, quality and reliability, is up to its author.
- GUI is not in the scope of this document.

> - Deploy clusters and components that a predefined checkup may require.

Reviewer: ```suggestion
- Deploying clusters
- Deploying components that a specific checkup may require.
```

Author: Done.

    - Once the checkup is finished, I can review the additional data, beyond the success/failure.
    - I can export the checkup results, cluster state and all other relevant information, so I can analyze them or pass
      them on to experts.

As a vendor:

- I want to write a checkup for a feature / service I provide, to validate the cluster that is aimed to use it.
    - I want to have a reference for a checkup, so it can help me write my own.
    - I want to use my technology of choice (e.g language, OS, etc).
    - I am able to run the checkups locally during development.
- I want to publish the checkups to my customers.
- I want to support my customers:
    - I want to ask the customer to run the checkups and collect the results.
    - I want to accept checkup results data from my customers, so I could analyze it.

As an integrator/support personnel:

- I want to support my customers:
    - Ask the customer to run the checkups and collect the results.
    - Accept checkup results data from my customers, so I could analyze it.
    - Forward checkup results to other experts (e.g. the vendor).

## API Extensions

### User Interface

#### Overview

> **Note**: The framework uses well-defined [ConfigMap](#configuration-and-output-specification) and [Job](#job-specification) objects (described below).

Author: Loved it, thanks.

1. Checkup launcher image.

## Phase 3 - Define a CRD with controllers and an operator
For this phase to happen, at least 7 checkups should adopt this framework.
### Required Functionality
1. The user should be able to deploy the framework using an operator.
2. The user should be able to configure, execute and watch the status of a checkup using a CR.
### Deliverables
1. CRD definition.
2. Controller(s) image.
3. Operator image + deployment manifests.

# Appendix A - Future User Stories
As a cluster administrator
- I want to check that my deployed cluster is correctly set up and is ready to work with my feature / service.
  - I want to be able to create a batch of checkups pending to run.
  - I want to be able to edit the batch prior to its run (add / remove checks).
  - I run a batch of checkups (the checkups will run in series / in parallel).
  - I run a checkup in an existing namespace. (renouncing the ability of the framework to make sure the checkup cleaned after itself)

# Appendix B - Future Improvements
Apply resourceQuata on the checkupNamespace - limit the number of resources.
Read artifacts from checkups.

# Appendix C - Design Decisions Considerations

> ## Exploration of Similar Projects

Author: Done.

Before designing this project, a thorough survey of similar projects was performed.
The results are depicted [here](./similar_projects_comparison.md).

> ## Permission Management

Author: Done.

> ## Permission Management

Author: `## Permission Management` is now right under `# Appendix C - Design Decisions Considerations`.

> **_Note_**: Kiagnose shall provide example manifest files for all the objects listed above.

For example: `kubectl delete -f /path/to/kiagnose.yaml`.

### Checkup Permissions Loading
#### Prerequisites
Cluster user / ServiceAccount with ‘create’ permissions to the following objects:
1. ClusterRole.
2. Role

#### Process Description
1. The **cluster administrator** shall **review** the permissions required by the checkup and **agree** to let a checkup use them.
2. The **cluster administrator** shall **create** the ClusterRole and Role objects required by the checkup.

### Kiagnose Execution
> **_Note_**: Kiagnose shall provide example manifest files for all the objects listed below.

#### Prerequisites
Cluster user / ServiceAccount with ‘create’ permissions the following objects:
1. A user-supplied ConfigMap object in the `kiagnose` namespace.
2. A Job object in Kiagnose’s namespace.

#### Process Description
1. The user shall create a ConfigMap object in the `kiagnose` namespace with all required input.
2. The user shall create a Job object of Kiagnose in the `kiagnose` namespace, which in turn shall run the concrete checkup.

### Checkup Setup
#### Prerequisites

> 1. Checkup Role and ClusterRole objects exist in the cluster.

Reviewer 2: Odd to see a single number.

   4.3. Create RoleBinding for the ServiceAccount for each Role object from the input.

5. Create a ClusterRoleBinding for the ServiceAccount for each ClusterRole object supplied from the input.
6. Create the checkup Job object in the checkup namespace, the underlying pod shall run with the previously created ServiceAccount.
7. Kiagnose shall record the checkup’s start timestamp at the results ConfigMap.

### Checkup Execution
1. Kiagnose shall monitor the checkup’s Job for successful completion, timeout or error codes, for example: not loading because it cannot get the checkup’s image.
2. Kiagnose shall record the checkup’s finish timestamp at the results ConfigMap.

### Checkup Teardown
#### Process Description
Kiagnose shall:
1. Copy the outputs from the results ConfigMap to the user-supplied ConfigMap object.
2. Delete the checkup Namespace object.
3. Delete all ClusterRoleBinding objects created at the ‘Checkup setup’ process.
4. Exit.

### Review Checkup Results
#### Prerequisites
The cluster user / ServiceAccount should have the permissions to list and read the following objects:

> 1. A user-supplied ConfigMap object in the `kiagnose` namespace.

Reviewer 2: Odd to see a single number.

Author: Done.

#### Process Description

> 1. The user shall read the user-supplied ConfigMap object they had previously created.

Reviewer 2: Odd to see a single number.

Author: Done.

### Checkup Results Removal
#### Prerequisites
The cluster user / ServiceAccount should have the permissions to list and read and delete the following objects:

> 1. A user-supplied ConfigMap object in the `kiagnose` namespace.

Author: Done.

- Third party vendors cannot use Kiagnose without registering themselves to it.

2. Opening a wide range of objects to read:

Cons:
- Can jeopardize cluster security without the admin’s consent.
- Kiagnose is responsible to manage this ability and to review and approve changes to it by the community.

## Running checkups in a dedicated namespace
It was required that Kiagnose made sure that checkups shall not litter the cluster with their objects.

Using a dedicated namespace for running a checkup, makes sure that the checkup has less possibility to harm the cluster, and that objects created by it are removed when the namespace is removed.

# Appendix D - Checkup Design Guidelines
The checkup shall:
1. Verify that all required inputs exists:
   - Results ConfigMap object.
   - Environment variables.
2. Get its current namespace.
3. Verify that all prerequisite components are available in the cluster (e.g Kubevirt, Multus, etc).
4. Set up the checkup objects in the checkup namespace.
5. Perform the checkup body.
6. Write the results to the results ConfigMap object.
7. Perform a teardown of checkup objects.
8. Exit.

# Appendix E - Open Issues

> 1. What to do with Role objects which are supposed to come from the user?

Author: Done.

- I want to check my deployed cluster with checkups that require my input to operate.
    - I pick the predefined checkup that checks a feature / service and run it.
    - The checkup will fail if I forget to specify an input value (e.g. ref to an existing NAD, namespace, etc).
    - The checkup will fail if my input is wrong (does not fit with what exists in the cluster).
    - In case the input validation fails, I can observe the status of the checkup as “setup-failed”.
    - In case my input is correct, the checkup should run like in the basic scenario.
    - I can export the checkup results, cluster state and all other relevant information, so I can analyze them or pass
      them on to experts.
- I want to check my deployed cluster with checkups that report additional data, beyond success/failure.
    - I pick the predefined checkup that checks a feature / service and run it.
    - The checkup should run like in the basic scenario.
    - Once the checkup is finished, I can review the additional data, beyond the success/failure.
    - I can export the checkup results, cluster state and all other relevant information, so I can analyze them or pass
      them on to experts.

> As a namespace administrator:

Reviewer: thanks, /lgtm

As a vendor:

- I want to write a checkup for a feature / service I provide, to validate the cluster that is aimed to use it.
    - I want to have a reference for a checkup, so it can help me write my own.
    - I want to use my technology of choice (e.g language, OS, etc).
    - I am able to run the checkups locally during development.
- I want to publish the checkups to my customers.
- I want to support my customers:
    - I want to ask the customer to run the checkups and collect the results.
    - I want to accept checkup results data from my customers, so I could analyze it.

As an integrator/support personnel:

- I want to support my customers:
    - Ask the customer to run the checkups and collect the results.
    - Accept checkup results data from my customers, so I could analyze it.
    - Forward checkup results to other experts (e.g. the vendor).

## Exploration of Similar Projects

Exploring existing solutions revealed several projects which aimed to run various tests on running clusters.
These solutions may be grouped into the following categories:

- Collection of tests that are grouped together under the same project with basic tooling to execute them.

> - Collection of tests that are grouped together for certification with tooling that help integrate their run and result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).

Reviewer 2: This line is too long. I would try to wrap lines to some reasonable length.
72, 100, whatever looks nice.

- Framework which can run arbitrary workloads ([Sonobuoy](https://sonobuoy.io/)).

>   The first two groups focus on self curated tests and not on arbitrary ones.

Reviewer 2: Newline is needed above.

Author: Done.

  In addition, the two groups are expected to run on dedicated (test) clusters and not on production clusters.
  These two points are at the core of our requirement and therefore not good candidates.

The closest solution found is Sonobuoy.
Based on the information gathered so far, its main limitation is the need to use a binary at the client side and the lack of a proper API (programmatic/declarative) without it.
Wrapping around this solution is an option, however, it brings the cost of maintaining this wrapper and still being limited by the underlying tool.

A table comparing the similar solutions can be found [here](./similar_projects_comparison.md).

## API Extensions

### User Interface

#### Overview

> **_NOTE:_** The framework uses well-defined [ConfigMap](#configuration-and-output-specification) and [Job](#job-specification) objects (described below).

1. The user shall create a **mutable** ConfigMap object in the `k8s-checkup-framework` namespace in order to:

- Provide input for the framework itself.
- Provide input for the selected checkup.
- Receive an output at the end of the checkup run.

> > **_NOTE:_** It shall be the user’s responsibility to remove the ConfigMap object after the results are read.

Author: We have it as a note under "Checkup Results Removal" section.

2. The user shall create a Job object in order to run the framework.
3. The user shall wait for the checkup Job to finish and get the checkup results from the ConfigMap object.

> > **_NOTE:_** This interface could evolve in the future to a CRD and a dedicated controller.

Author: Done.

|--------------------------------------|------------------------------------------------------------------------------------------|--------------|------------------------------------------------------------------------------|
| `spec.image`                         | Checkup image                                                                            | True         |                                                                              |
| `spec.timeout`                       | Timeout to wait for a checkup to finish [min]                                            | True         |                                                                              |
| `spec.param.<checkup argument name>` | Arbitrary parameter required by the concrete checkup                                     | False        | 0..N                                                                         |
| `spec.clusterRoles`                  | ClusterRole required by the checkup Value format: `<ClusterRole name>` Newline separated | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |

> | `spec.roles`                         | Role required by the checkup Value format: `<namespace/name>` Newline separated          | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |

Reviewer 2: Can some columns be shrink by splitting the text with a newline?
It may look fine on a rendered page, but to review this it is hard.

Authors of features and services should be able to easily write and publish predefined checkups that will verify that
their feature / service works as expected on a pre-deployed cluster.

## Goals

- Expose a checkup to a cluster.
- Run the predefined checkup on a cluster.
- Provide a reference implementation of a checkup.
- Do not disrupt or interfere with existing cluster workloads. This includes making sure no leftovers exist on
  termination.
- Checkup execution should require minimal dependencies, allowing a cluster administrator or some management tool to
  trigger it.
- Deploy and run the checkups on any K8s-based cluster.
- Provide a well-defined API, allowing management components to interact with the framework.

## Non Goals

- Defining content of the checkup, including its correctness, quality and reliability, is up to its author.
- GUI is not in the scope of this document.
- Deploy clusters and components that a predefined checkup may require.

# Proposal

## Definition of Personas

- K8s cluster administrators.
- Vendors who provide verification of their K8s extensions.
- Installation, integration & support teams, which examine cluster functionality.

## User Stories

> As a cluster administrator:

Reviewer: Can we somehow refine this? I can think of another a non-admin, who got access to a cluster and now would like to check if the cluster fits their needs.

I think that it is important that a non-privileged user would be able to run the checkups. If a specific checkup fails due to lack of permissions, that *exactly* what the user needs to know.

Author: @dankenigsberg, we consider a checkup to be an untrusted arbitrary code running inside a cluster.
The framework can bind existing ClusterRole objects with the checkup's ServiceAccount - the user only have provide a ClusterRole name and it is automatically bound.
In other words, if you know a ClusterRole's name - you can use it.

From a security standpoint, IMO the users that could configure and run checkups should be trusted, because they wield a great power.

In order to configure and run a checkup, a user is required to have the following permissions in the framework's namespace:
1. Read/write ConfigMaps (or a CR in the near future).
2. Create a Job.

So full admin privileges are not required in order to use the framework.

Reviewer: I think that the use case of non-privileged tester should be written in this document even if the current implementation does not support it. If I was a cluster admin, I would *not* run some random checkup container as an admin. I'd prefer to have a clear  request on which namespace I need to create, which permissions should I grant to the namespace admin, and only then run the checkup.

Reviewer 2: This was not a goal or requirement.
Kiagnose is privileged and trusted, with elevated permissions (e.g. create namespaces for the checkup workload).
On the other hand, the checkup themselves are untrusted and requires the Admin to register their permissions explicitly before usage.

When (and if) kiagnose will start using a controller with a CR, we would probably be able to give permissions to who can create such CR/s and by that allow different users to execute checkups.

Reviewer 2: @dankenigsberg , would you like to add to this thread?
We would like to merge this design proposal before the first release tag, so I would like to make sure we also finalize this topic.

Reviewer: > would you like to add to this thread?

Sorry, I missed your month-old questions.

> This was not a goal or requirement.

That's why a proposal review is important.

I have an OpenShift cluster with a very strict admin. Only he can create projects (=namespaces) and set their admin users. I can ask him to grant special rights to a project of mine (e.g more quota - the default is very very small, assign a specific NAD to it etc) but he would outright reject installing an opensource deployment with unlimited permissions.

He would be fine if I run the checkup on *my* project, consuming *my* quota, testing the specific NAD allocated to me, under *my* privileges.

I'm not asking to do it *now*, but to me, it is important that the framework has a lightweight mode which expects all resources to be prepared ahead of time. I don't mind if it also has another module that attempts to create the test namespace and clean it after the fact.

Reviewer: > I agree that the framework has a lot of power, which could lead to security problems.

It's not just that. The cluster I talk about has annoyingly low default quota. Unless the framework can outguess the quota required by the individual checkup, it would not run properly. And if it does whatever the checkup declare it needs, it would be easy for a malicious checkup to starve the system.

> I think we should discuss what are the requirements and limitations when running a checkup in an existing namespace.

+1

> One obvious limitation will probably be that the framework could not enforce a proper cleanup of all objects created by a checkup.

I'm ok with documenting that a  buggy/malicious checkup may leave content in my namespace, which I'm advised to empty after usage.

> a checkup (or the framework) could not run under a user's privileges

I can run a pod and a VM under my human privileges, why not a checkup?

If I must make the service account an admin of my namespace before running the checkup, so be it. My cluster-admin would not be happy if this account is too privileged, though.

    - I can export the checkup results, cluster state and all other relevant information, so I can analyze them or pass
      them on to experts.

As a vendor:

- I want to write a checkup for a feature / service I provide, to validate the cluster that is aimed to use it.
    - I want to have a reference for a checkup, so it can help me write my own.
    - I want to use my technology of choice (e.g language, OS, etc).
    - I am able to run the checkups locally during development.
- I want to publish the checkups to my customers.
- I want to support my customers:
    - I want to ask the customer to run the checkups and collect the results.
    - I want to accept checkup results data from my customers, so I could analyze it.

As an integrator/support personnel:

- I want to support my customers:
    - Ask the customer to run the checkups and collect the results.
    - Accept checkup results data from my customers, so I could analyze it.
    - Forward checkup results to other experts (e.g. the vendor).

## Exploration of Similar Projects

Exploring existing solutions revealed several projects which aimed to run various tests on running clusters.

> These solutions may be grouped into the following categories:

Reviewer 3: This line wont start at new line.

Author: Done.

> These solutions may be grouped into the following categories:

Reviewer 3: This line wont start at new line.

Author: Is this comment a duplicate of the previous comment?

- Collection of tests that are grouped together under the same project with basic tooling to execute them.
- Collection of tests that are grouped together for certification with tooling that help integrate their run and
result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).
- Framework which can run arbitrary workloads ([Sonobuoy](https://sonobuoy.io/)).

> The first two groups focus on self curated tests and not on arbitrary ones.

Reviewer 3: Typo  "The first two groups focus on self curated tests and not **on** arbitrary ones.", it should be "an".

Author: This is not a typo, I've verified it with a [grammar checking tool](https://languagetool.org/).

> The first two groups focus on self curated tests and not on arbitrary ones.

Reviewer 3: Typo  "The first two groups focus on self curated tests and not **on** arbitrary ones.", it should be "an".

Author: Is this comment a duplicate of the previous one?

In addition, the two groups are expected to run on dedicated (test) clusters and not on production clusters.
These two points are at the core of our requirement and therefore not good candidates.

The closest solution found is Sonobuoy.
Based on the information gathered so far, its main limitation is the need to use a binary at the client side and the
lack of a proper API (programmatic/declarative) without it.
Wrapping around this solution is an option, however, it brings the cost of maintaining this wrapper and still being
limited by the underlying tool.

A table comparing the similar solutions can be found [here](./similar_projects_comparison.md).

## API Extensions

### User Interface

#### Overview

> **_NOTE:_** Kiagnose uses well-defined [ConfigMap](#configuration-and-output-specification) and
> [Job](#job-specification) objects (described below).
> 
> This interface could evolve in the future to a 
> [CRD](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) and a dedicated controller.

> 1. The user shall create a **mutable** ConfigMap object in the `k8s-checkup-framework` namespace in order to:

Author: Done all over.

| `spec.image`                         | Checkup image                                                                            | True         |                                                                              |
| `spec.timeout`                       | Timeout to wait for a checkup to finish [min]                                            | True         |                                                                              |
| `spec.param.<checkup argument name>` | Arbitrary parameter required by the concrete checkup                                     | False        | 0..N                                                                         |
| `spec.clusterRoles`                  | ClusterRole required by the checkup Value format: `<ClusterRole name>` Newline separated | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |
| `spec.roles`                         | Role required by the checkup Value format: `<namespace/name>` Newline separated          | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |

##### Output Fields

> 1. The user shall expect the checkup results concatenated to the ConfigMap object at the end of the checkup run.

Author: Here it is actually the user-supplied ConfigMap object.
I've made it explicit.

      env:
      - name: CONFIGMAP_NAMESPACE
        value: <ConfigMap namespace>
      - name: CONFIGMAP_NAME
        value: <ConfigMap name>
...
```

#### Checkup Execution Example
##### Input
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: echo-checkup-example-config
  namespace: k8s-checkup-framework
data:
  spec.image: registry:5000/echo-checkup:latest
  spec.timeout: 1m
  spec.param.message: "Hi!"
---
apiVersion: batch/v1
kind: Job
metadata:
  name: checkup-launcher
  namespace: k8s-checkup-framework
spec:
  backoffLimit: 0
  template:
    spec:
      serviceAccount: checkup-launcher-sa
      restartPolicy: Never
      containers:
        - name: checkup-launcher
          image: registry:5000/checkup-launcher:latest
          env:
          - name: CONFIGMAP_NAMESPACE
            value: k8s-checkup-framework
          - name: CONFIGMAP_NAME
            value: echo-checkup-example-config
...
```

##### Minimal Checkup Entrypoint
```bash
#!/bin/sh
OUTPUT_FILE="/tmp/result_configmap.yaml"

cat <<EOF >$OUTPUT_FILE
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: $RESULT_CONFIGMAP_NAME
  namespace: $RESULT_CONFIGMAP_NAMESPACE
data:
  status.succeeded: "true"

>   status.failureReason: none

Author: Done.

  status.result.echo: "$MESSAGE"
...
EOF

kubectl patch configmap "$RESULT_CONFIGMAP_NAME" -n "$RESULT_CONFIGMAP_NAMESPACE" --patch-file $OUTPUT_FILE
```

##### Output
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: echo-checkup-example-config
  namespace: k8s-checkup-framework
data:
  spec.image: registry:5000/echo-checkup:latest
  spec.timeout: 1m
  spec.param.message: "Hi!"
  status.succeeded: “true”

>   status.failureReason: none

Author: Done.

> Kiagnose - Kubernetes Checkup Framework

Reviewer: Regarding the diagram, it could be even cooler if it was in svg (which smaller to save and easier to modify)

Reviewer 2: @orelmisan , could you please respond?

Author: Done.

=

> **Authors**: Edward Haas, Orel Misan, Or Mergi

Reviewer: care to add your @orelmisan for "bragging rights"?

Author: Done.

# Summary

> This proposal introduces a framework for running K8s cluster checkups, which verifies that a certain cluster
> functionality is working as expected.

Reviewer: ```suggestion
This proposal introduces a framework to verify that certain cluster
functionalities are working as expected.
```

I think that at this stage, the reader does not yet know what is a "cluster checkup", so a shorter summary may be better.

Author: Done.

# Motivation

> Operating a K8s cluster is challenging. Such clusters usually provide a wide range of functionalities and services, some

Reviewer: ```suggestion
Operating a [K8s](https://kubernetes.io/) cluster is challenging. Such clusters usually provide a wide range of functionalities and services, some
```
in case your reader does not know what k8s is.

Author: Done.

provided by K8s itself, some by downstream distribution of it (e.g. OpenShift) and others by third-party vendors. All
are expected to co-exist and work properly on various deployments (on cloud providers, bare-metal).

Clusters that provide multiple functionalities are challenging and time-consuming for their administrators to
troubleshoot. Understanding why a specific functionality is not working as expected requires deep understanding of how
it works, its interaction and dependencies with other parts of the cluster. Considering that each deployment is unique,
it is time-consuming for both the cluster administrator and the vendor of a functionality to troubleshoot unexpected
behavior.

To simplify cluster operation and maintenance, it will be beneficial to allow cluster administrators to validate that
the cluster can support a specific functionality/service and to provide a simple indication that they are working as
intended.

Authors of features and services should be able to easily write and publish predefined checkups that will verify that
their feature / service works as expected on a pre-deployed cluster.

## Goals

> - Expose a checkup to a cluster.

Reviewer: I don't understand this goal

Reviewer 2: The cluster needs to know the checkup is there.
At the minimum, we need to tell someone where it is and it should be accessible from the cluster.

Originally we thought about it as "register checkup to the cluster" but there was some resistance and it was too specific.

Reviewer: I still do not understand it... The checkup framework that you suggest here is responsible to expose *a* checkup? Or is it about exposing multiple checkups, some of them predefined, some of them user-defined, some of them by third parties?

Author: @EdDev @dankenigsberg WDYT about "Deploy a checkup to a cluster"?

- Run the predefined checkup on a cluster.
- Provide a reference implementation of a checkup.
- Do not disrupt or interfere with existing cluster workloads. This includes making sure no leftovers exist on
  termination.
- Checkup execution should require minimal dependencies, allowing a cluster administrator or some management tool to
  trigger it.

> - Deploy and run the checkups on any K8s based cluster.

Reviewer: ```suggestion
- Deploy and run the checkups on any K8s-based cluster.
```

Author: Done.

> - Provide a well-defined API, allowing management components to interact with the solution.

Reviewer: ```suggestion
- Provide a well-defined API, allowing management components to interact with the framwork.
```

I prefer using the same term for what you build here. (unless you mean something else by "solution")

## Non Goals

- Defining content of the checkup, including its correctness, quality and reliability, is up to its author.
- GUI is not in the scope of this document.

> - Deploy clusters and components that a predefined checkup may require.

Reviewer: I don't understand the English here.

Reviewer 2: It tries to say that dependencies the checkup requires are not going to be deployed or managed by kiagnose.

Reviewer 2: @orelmisan , would you want to try and rephrase it?

Author: @EdDev @dankenigsberg WDYT about "Deploying checkup's dependencies"?

# Proposal

> ## Definition of Users

Reviewer: ```suggestion
## Definition of Personas
```
I find it as a clearer term.

Author: Done.

- K8s cluster administrators.
- Vendors who provide verification of their K8s extensions.

> - Installation, integration & support teams, which examine cluster functionality.

Reviewer: How is installation team different from cluster admin?
Can you focus the third persona on a single support engineer? I find it confusing to consider three types of teams as a persona.

Reviewer 2: > How is installation team different from cluster admin?

One installs the cluster, possibly for the first time.
While the Admin may just maintain it afterwards.

Reviewer: I don't see how the differentiation between the two helps the discussion here. Also I don't understand how the installation team is related to the proposed framework, and why they should be bundled with the support team. I still think that the persona definition would be improved if you focus on one of them.

Author: The intent was "people which installs a cluster and people which support existing clusters".

Reviewer: I don't see how "people which install a cluster" is an interesting persona here, nor how its behavior and requirement is different from that of the cluster admin.

Author: We can change these personas to:
- cluster admins (and all the other personas with similar permissions).
- namespace admins.
- vendors.

WDYT?
